src/main/java/eu/europeana/downloads/ListIdentifierExecutor.java

```diff
@@ -0,0 +1,118 @@
+    private String directoryLocation;
+    private int logProgressInterval;
+
+    public ListIdentifierExecutor(List<String> sets, String from, String metadataPrefix, String directoryLocation, OAIPMHServiceClient oaipmhServer, int logProgressInterval) {
+        this.sets = sets;
+        this.from = from;
+        this.metadataPrefix = metadataPrefix;
+        this.directoryLocation = directoryLocation;
+        this.oaipmhServer = oaipmhServer;
+        this.logProgressInterval = logProgressInterval;
+    }
+
+    @Override
+    public ListRecordsResult call() {
+        int errors = 0;
+        long counter = 0;
+        long start = System.currentTimeMillis();
+
+//         This is a bit of a hack. The first callable that reaches this point will create a progressLogger and only
+//         this callable will log progress. This is to avoid too much logging from all threads.
+        synchronized (this) {
+            if (logger == null) {
+                logger = new ProgressLogger("ListIdentifiers Sets", sets.size(), logProgressInterval);
+                loggerThreadId = Thread.currentThread().getId();
+                LOG.debug("Created new progress logger for thread {} - {} items, logging interval {} ms",
+                        loggerThreadId, sets.size(), logProgressInterval);
+            }
+        }
+        for (String set : sets) {
+            try {
+                ListIdentifiersQuery identifiersQuery = new ListIdentifiersQuery(metadataPrefix, from, set, logProgressInterval);
+                identifiersQuery.initSets();
+                List<String> identifiers = identifiersQuery.getIdentifiers(oaipmhServer, set);
+                // start downloading the collected identifiers
+                errors += harvestIdentifiers(identifiers, set);
+            } catch (Exception e) {
+                LOG.error("Error collecting identifiers for set {}", set, e);
+                errors++;
+            }
+            // if there are too many errors, just abort
+            if (errors > MAX_ERRORS_PER_THREAD) {
+                LOG.error("Terminating getIdentifiers thread {} because too many errors occurred", Thread.currentThread().getId());
+                break;
+            }
+            if (Thread.currentThread().getId() == loggerThreadId) {
+                counter++;
+                logger.logProgress(counter);
+            }
+
+        }
+        return new ListRecordsResult((System.currentTimeMillis() - start) / 1000F, errors);
+    }
+
+    /**
+     * Method to harvest identifiers retreived for each set. Creates a zip folder with the setname
+     * in the specified location and generates md5 sum file for the zip.
+     *
+     * @param identifiers list of identifiers
+     * @param setIdentifier setName
+     * @return number of errors occurred during the process to terminate the thread in progress.
+     */
+    private int harvestIdentifiers(List<String> identifiers, String setIdentifier) {
```

Comment:
I am a bit confused about the description; it says it harvests identifiers, but looking at the code it seems you already have the identifiers and are actually harvesting the actual records (retrieved via getRecord)?

Reply:
Yes, I will change the description here.